src/cpp/transport/serial/MultiSerialAgentLinux.cpp

```diff
@@ -0,0 +1,224 @@
+                            UXR_DECORATE_YELLOW("[==>> SER <<==]"),
+                            raw_client_key,
+                            input_packet.message->get_buf(),
+                            input_packet.message->get_len());
+                    }
+                }
+            }
+        }
+    }
+    else
+    {
+        transport_rc = (ret == 0) ? TransportRc::timeout_error : TransportRc::server_error;
+    }
+
+    return rv;
+}
+
+bool MultiSerialAgent::send_message(
+        OutputPacket<MultiSerialEndPoint> output_packet,
+        TransportRc& transport_rc)
+{
+    bool rv = false;
+    int client_fd = output_packet.destination.get_fd();
+
+    utils::SharedLock lk(framing_mtx);
+    std::map<int, FramingIO>::iterator it = framing_io.find(client_fd);
+    
+    if (it == framing_io.end())
+    {
+        // Destination client not found on active ports
+        return rv;
+    }
+
+    ssize_t bytes_written =
+            it->second.write_framed_msg(
+                output_packet.message->get_buf(),
+                output_packet.message->get_len(),
+                output_packet.destination.get_addr(),
+                transport_rc);
+
+    if ((0 < bytes_written) && (
+         static_cast<size_t>(bytes_written) == output_packet.message->get_len()))
+    {
+        rv = true;
+
+        uint32_t raw_client_key;
+        if (Server<MultiSerialEndPoint>::get_client_key(output_packet.destination, raw_client_key))
+        {
+            UXR_AGENT_LOG_MESSAGE(
+                UXR_DECORATE_YELLOW("[** <<SER>> **]"),
+                raw_client_key,
+                output_packet.message->get_buf(),
+                output_packet.message->get_len());
+        }
+    }
+    return rv;
+}
+
+ssize_t MultiSerialAgent::read_data(
+        uint8_t serial_fd,
+        uint8_t* buf,
+        size_t len,
+        int timeout,
+        TransportRc& transport_rc)
+{
+    // Modify timeout?
+    ssize_t bytes_read = 0;
+    pollfd read_file = {serial_fd, POLLIN, 0};
+    
+    int poll_rv = poll(&read_file, 1, timeout);
+
+    if(read_file.revents & (POLLERR+POLLHUP))
+    {
+        transport_rc = TransportRc::server_error;;
+    }
+    else if (0 < poll_rv)
+    {
+        bytes_read = read(serial_fd, buf, len);
+        if (0 > bytes_read)
+        {
+            transport_rc = TransportRc::server_error;
+        }
+    }
+    else
+    {
+        transport_rc = (poll_rv == 0) ? TransportRc::timeout_error : TransportRc::server_error;
+    } 
+    
+    if (transport_rc == TransportRc::server_error)
+    {
+        std::unique_lock<std::mutex> lk(error_mtx);
+        error_fd.push_back(serial_fd);
+    }
+    
+    return bytes_read;
+}
+
+ssize_t MultiSerialAgent::write_data(
+        uint8_t serial_fd,
+        uint8_t* buf,
+        size_t len,
+        TransportRc& transport_rc)
+{
+    size_t rv = 0;
+    ssize_t bytes_written = ::write(serial_fd, buf, len);
+    if ((0 < bytes_written)  && size_t(bytes_written) == len)
```

Comment:
Check `FramingIO::transport_write``I think that this function can send less bytes than `len`

Reply:
True (write_buffer_pos_ is len parameter):
https://github.com/eProsima/Micro-XRCE-DDS-Agent/blob/ed66bb0ffecbb1f234fefb21ee9af4a3e2c7cb86/src/cpp/transport/stream_framing/StreamFramingProtocol.cpp#L397-L402

Modified on Serial and Multiserial modes: cd18d34